### DISCUSSION TITLE
Grifeing Attack

### DISCUSSION THREAD

**Discussion by kalpshah284** [clearpool-finance/community#14] [Bug reports]
Bug Description

- Where:
     `provideWithPermit` and `provideForWithPermit` 

- Expected behavior:
     The  `provideWithPermit` and `provideForWithPermit` functions utilize the permit function so that approve and pull operations can happen in a single transaction instead of two consecutive transactions.

- Attack:

   ERC20Permit uses the nonces mapping for replay protection. Once a signature is verified and approved, the nonce increases, invalidating the same signature being replayed.

   provideWithPermit and provideForWithPermit expect the holder to sign their tokens and provide the signature to the contract as part of `permitData`

    When a provideWithPermit and provideForWithPermit transaction is in the mempool, an attacker can take this signature, and call the `permit` function on the token themselves.

    Since this is a valid signature, the token accepts it and increases the nonce.

    This makes the spender's transaction fail whenever it gets mined. 

-Impact

An attacker can ensure all calls to provideWithPermit and provideForWithPermit fail for the first time. 

-Risk Breakdown

 Difficulty to Exploit: Easy
 Weakness: 
 CVSS2 Score:

-Recommendation

 In the provideWithPermit and provideForWithPermit function, check if it has the approval it needs. If not, then only submit the permit signature.
    ```
        if (IERC20(_token).allowance(owner, spender) < amount) {
            IERC20PermitUpgradeable(_token).permit(msg.sender, address(this), amount, deadline, v, r, s);
        }
    ```

**Comment by georgeciubotaru** [comment]
Hi @kalpshah284 thank you for reporting the issue. I've passed the issue on to our security team. I will let you know the resolution.

**Reply by kalpshah284** [reply]
thank you sir

**Reply by zolotokrylin** [reply]
@kalpshah284 thank you for the report. Can you please send me a message in Discord.`zolotokrylin` – my user name.

**Reply by kalpshah284** [reply]
yes sir

**Reply by kalpshah284** [reply]
I send request `kalp284`
